### ISSUE TITLE
terraform apply - error unknown provider

### ISSUE THREAD

**Issue by melbit-naveennathan** [hashicorp/terraform#20191]
### Terraform Version
```
$ terraform -v
Terraform v0.11.11
+ provider.aws v1.56.0
+ provider.template v1.0.0
```

### Debug Output

https://gist.github.com/melbit-naveennathan/1660bc2daa3d86f3f85a99913cf4f3c9

### Expected Behavior
terraform apply should work

### Actual Behavior

I get the following errors:

```
Error asking for user input: 4 error(s) occurred:

* provider.tier-app-epgapi: unknown provider "tier-app-epgapi"
* provider.tier-app-bau: unknown provider "tier-app-bau"
* provider.tier-app-epg: unknown provider "tier-app-epg"
* provider.tier-app-epgweb: unknown provider "tier-app-epgweb"
```

### Steps to Reproduce
```
rm -rf .terraform
terraform init
terraform apply
```

### Additional Context

It seems the `terraform state pull` operation works. It just seems to have issues when doing terraform apply.

Note there is no provider called "tier-app-...". There are however module calls that use this:

```
$ grep module tier.tf
module "tier-app-bau" "app-bau" {
module "tier-app-wrapper" "app-wrapper" {
module "tier-app-pram" "app-pram" {
module "tier-app-epgapi" "app-epgapi" {
module "tier-app-epgweb" "app-epgweb" {
module "tier-app-epg" "app-epg" {
module "tier-app-onlineapigw" "app-onlineapigw" {
module "tier-app-myaccount" "app-myaccount" {
```

Somehow terraform seems confused.

**Comment by melbit-naveennathan** [comment]
Thanks for the tip about module. I've removed the excess label but the same issue persists.

The terraform config can be [downloaded here](https://user.fm/files/v2-a553a8393ebb05ad4384aa5eb70a076f/terraform.tar.gz.gpg) encrypted under the security@hashicorp.com gpg key.

Since there's a few modules being called, I just uploaded the terraform portion that we do `terraform apply` including the `.terraform` directory (minus plugins to keep the size of the tarball small).

Thanks.

**Comment by melbit-naveennathan** [comment]
So I don't think the existing configuration is the issue here.

I removed the `main.tf` which contained the bulk of the terraform code (i.e. calls external modules and creates resources).

I think the issue is specifically with the state file.

```
src2 nn$ rm -rf .terraform main.tf
src2 nn$ terraform init

Initializing the backend...

Successfully configured the backend "s3"! Terraform will automatically
use this backend unless the backend configuration changes.

Initializing provider plugins...
- Checking for available provider plugins on https://releases.hashicorp.com...
- Downloading plugin for provider "template" (1.0.0)...
- Downloading plugin for provider "aws" (1.57.0)...

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
src2 nn$ terraform get
src2 nn$ ls -la .terraform/modules
ls: .terraform/modules: No such file or directory
src2 nn$ terraform plan

Error: Error asking for user input: 4 error(s) occurred:

* provider.tier-app-epg: unknown provider "tier-app-epg"
* provider.tier-app-epgweb: unknown provider "tier-app-epgweb"
* provider.tier-app-bau: unknown provider "tier-app-bau"
* provider.tier-app-epgapi: unknown provider "tier-app-epgapi"

src2 nn$ cat *.tf
provider "aws" {
  region  = "${var.region}"
  version = "~> 1.32"
}

terraform {
  required_version = "~> 0.11"

  backend "s3" {
    bucket         = "ft-terraform-state"
    key            = "ft/sit.tfstate"
    encrypt        = "true"
    region         = "ap-southeast-2"
    dynamodb_table = "terraform-lock"
  }
}

variable "region" {
  default     = "ap-southeast-2"
  description = "AWS Access Region"
}
provider "template" {
  version = "~> 1.0"
}
```

**Comment by melbit-naveennathan** [comment]
I've fixed the issue on my end. I deleted the resources from the _root_ module in the state file, specifically:
 - tier-app-bau.aws_iam_role
 - tier-app-epg.aws_iam_role
 - tier-app-epgapi.aws_iam_role
 - tier-app-epgweb.aws_iam_role

I then deleted the corresponding resources in AWS.

The actual module "tier-app-{bau,epg,epgapi,epgweb}" were left untouched.

This has fixed the issue.

**Comment by apparentlymart** [comment]
It seems like you somehow ended up with some invalid-named resources in your state there, which Terraform then misunderstood as being resource types, and in turn provider types.

Did you use any of the `terraform state ...` family of commands on this state prior to the problem occurring? They are supposed to validate given input but Terraform's terse address syntax can sometimes lead to a mistyped address being misunderstood rather than recognized as invalid. Possibly you have found a place where the validation isn't precise enough.

**Comment by apparentlymart** [comment]
Great, thanks for confirming, @melbit-naveennathan! It sounds like there is some missing validation in `terraform state mv` then, which (presumably) allowed you to change the resource type as part of a `mv`, which should be blocked since that command is for renaming resources and moving them between modules.

Fortunately it looks like we already added a check for that as part of the 0.12 development cycle:

https://github.com/hashicorp/terraform/blob/d1daa7e5187d0120620f260ec1aed9631b77f6f8/command/state_mv.go#L172-L176

...so we'll use this issue to prompt us to verify that fix as part of preparing the 0.12.0 release.

**Comment by jbardin** [comment]
Verified that 0.12 now blocks the invalid state operations.
Thanks @melbit-naveennathan!
